`x:Bind` casting not working with `muxc:NavigationViewItem`

### Current behavior

Not sure if just missed scenarios with #4855?

_Both of these scenarios work in UWP_:

```xml
<controls:SwitchPresenter Value="{x:Bind ((muxc:NavigationViewItem)NavView.SelectedItem).Name, Mode=OneWay}">
```

```xml
<controls:SwitchPresenter Value="{x:Bind NavView.SelectedItem.(muxc:NavigationViewItem.Name), Mode=OneWay}">
```

As called out as valid casting methods [in the docs](https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension#casting)

However, they both cause invalid generated bindings when compiling on WASM/WPF:

```cs
c14.SetBinding(global::Microsoft.Toolkit.Uwp.UI.Controls.SwitchPresenter.ValueProperty, new Windows.UI.Xaml.Data.Binding{ Mode = global::Windows.UI.Xaml.Data.BindingMode.OneWay/* Windows.UI.Xaml.Data.BindingMode/Windows.UI.Xaml.Data.BindingMode, OneWay, Bind/Mode */,  }.BindingApply(___b =>  /*defaultBindModeOneTime ((global::Microsoft.UI.Xaml.Controls.NavigationViewItem)NavView.SelectedItem).Name*/ global::Uno.UI.Xaml.BindingHelper.SetBindingXBindProvider(___b, this, ___ctx => ___ctx is global::MyApp.MainPage ___tctx ? (object)(___tctx.((global::Microsoft.UI.Xaml.Controls.NavigationViewItem)NavView.SelectedItem).Name) : null, null , new [] {"((global::Microsoft.UI.Xaml.Controls.NavigationViewItem)NavView.SelectedItem).Name"})));
```

`error CS1001: Identifier expected`

And for the second one:

```cs
c14.SetBinding(global::Microsoft.Toolkit.Uwp.UI.Controls.SwitchPresenter.ValueProperty, new Windows.UI.Xaml.Data.Binding{ Mode = global::Windows.UI.Xaml.Data.BindingMode.OneWay/* Windows.UI.Xaml.Data.BindingMode/Windows.UI.Xaml.Data.BindingMode, OneWay, Bind/Mode */,  }.BindingApply(___b =>  /*defaultBindModeOneTime NavView.SelectedItem.(global::Microsoft.UI.Xaml.Controls.NavigationViewItem.Name)*/ global::Uno.UI.Xaml.BindingHelper.SetBindingXBindProvider(___b, this, ___ctx => ___ctx is global::MyApp.MainPage ___tctx ? (object)(___tctx.NavView.SelectedItem.(___tctx.global::Microsoft.UI.Xaml.Controls.NavigationViewItem.Name)) : null, null , new [] {"NavView.SelectedItem"})));
```

with `error CS1001: Identifier expected` and `error CS7000: Unexpected use of an aliased name`

### Expected behavior

Able to use a cast to bind to a NavigationViewItem property with x:Bind.

### How to reproduce it (as minimally and precisely as possible)

1. Create an uno project.
2. Add the Toolkit Primitives nugets
3. Create a NavigationView/SwitchPresenter setup:
```xml
     <muxc:NavigationView x:Name="NavView"
                         PaneDisplayMode="Top" IsBackButtonVisible="Collapsed">
        <muxc:NavigationView.MenuItems>
            <muxc:NavigationViewItem x:Name="ItemHelp">
                <muxc:NavigationViewItem.Icon>
                    <FontIcon FontFamily="{ThemeResource SymbolThemeFontFamily}" Glyph="&#xE9CE;" />
                </muxc:NavigationViewItem.Icon>
            </muxc:NavigationViewItem>
            <muxc:NavigationViewItem x:Name="ItemOther" Content="Other" Icon="Calendar"
                                     IsSelected="True"/>
            <muxc:NavigationViewItem x:Name="ItemOther2" Content="Other 2">
                <muxc:NavigationViewItem.Icon>
                    <FontIcon FontFamily="{ThemeResource SymbolThemeFontFamily}" Glyph="&#xF272;" />
                </muxc:NavigationViewItem.Icon>
            </muxc:NavigationViewItem>
        </muxc:NavigationView.MenuItems>
        <controls:SwitchPresenter Value="{x:Bind NavView.SelectedItem.(muxc:NavigationViewItem.Name), Mode=OneWay}">
            <controls:Case Value="ItemHelp">
                <local:HelpPage/>
            </controls:Case>
            <controls:Case Value="ItemOther">
                <local:OtherPage/>
            </controls:Case>
            <controls:Case Value="ItemOther2">
                <local:OtherPage2/>
            </controls:Case>
        </controls:SwitchPresenter>
    </muxc:NavigationView>
```
Use either of the binding setups provided at the top for the SwitchPresenter.
4. Run and see UWP app runs and works.
5. Try compiling for WASM/WPF, get compile error.

### Workaround

Use a regular binding expression:
```xml
<controls:SwitchPresenter Value="{Binding SelectedItem.Name, ElementName=NavView}">
```

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

4.2.0-dev.156

### Affected platforms

WebAssembly, Skia (WPF)

### IDE

Visual Studio 2022

### IDE version

17.1.0

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

I've tested this on Windows, before dev.1646 and after.

```
{x:Bind ((muxc:NavigationViewItem)NavView.SelectedItem).Name, Mode=OneWay}
```

Windows: works
Pre 1646: don't compile: error CS1001: Identifier expected
Post 1646: compiles, but doesn't give any value (e.g. I tested with a textblock and the textblock is blank)


```
{x:Bind NavView.SelectedItem.(muxc:NavigationViewItem.Name), Mode=OneWay}
```

Windows: works
Pre 1646: don't compile: error CS1001: Identifier expected
Post 1646: don't compile: error CS1001: Identifier expected

--- a/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/XBindExpressionParser.cs
+++ b/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/XBindExpressionParser.cs
@@ -55,11 +55,11 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 				_findType = findType;
 			}
 
-			private int FirstIndexOf(char c)
+			private static int LastIndexOf(StringBuilder builder, char c)
 			{
-				for (int i = 0; i < _builder.Length; i++)
+				for (int i = builder.Length - 1; i >= 0; i--)
 				{
-					if (_builder[i] == c)
+					if (builder[i] == c)
 					{
 						return i;
 					}
@@ -68,72 +68,56 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 				return -1;
 			}
 
-			private int LastIndexOf(char c)
+			private static void AddIfNotStaticAndClear(StringBuilder propertyBuilder, ImmutableArray<string>.Builder propertiesBuilder)
 			{
-				for (int i = _builder.Length - 1; i >= 0; i--)
+				if (propertyBuilder.Length == 0)
 				{
-					if (_builder[i] == c)
-					{
-						return i;
-					}
+					// Only add to propertiesBuilder if we have something in propertyBuilder.
+					return;
 				}
 
-				return -1;
-			}
-
-			private bool StartsWith(string s, int startIndexToSearchFrom = 0)
-			{
-				if (startIndexToSearchFrom + s.Length > _builder.Length)
+				var property = propertyBuilder.ToString();
+				if (!property.StartsWith("global::", StringComparison.Ordinal) &&
+					property is not ("null" or "true" or "false"))
 				{
-					return false;
+					propertiesBuilder.Add(property);
 				}
 
-				for (int i = 0; i < s.Length; i++)
-				{
-					if (_builder[startIndexToSearchFrom++] != s[i])
-					{
-						return false;
-					}
-				}
-
-				return true;
+				propertyBuilder.Clear();
 			}
 
 			public (string CSharpExpression, ImmutableArray<string> Properties) Build(XBindRoot root)
 			{
+				var propertiesBuilder = ImmutableArray.CreateBuilder<string>();
+				var propertyBuilder = new StringBuilder();
 				if (root is XBindInvocation invocation)
 				{
-					var propertiesBuilder = ImmutableArray.CreateBuilder<string>();
+					BuildPath(invocation.Path, propertyBuilder);
 
-					BuildPath(invocation.Path);
-
-					// If we have an invocation on the form of context.SomePath.Method(...)
+					// If we have an invocation on the form of SomePath.Method(...)
 					// Then we add "SomePath" to the propertiesBuilder
-					var firstIndexOf = FirstIndexOf('.');
-					var lastIndexOf = LastIndexOf('.');
-					if (firstIndexOf != lastIndexOf && StartsWith(_contextName))
+					// Note that the _builder field will contain the contextName followed by '.' at the beginning,
+					// while propertyBuilder doesn't.
+					var lastIndexOf = LastIndexOf(propertyBuilder, '.');
+					if (lastIndexOf != -1)
 					{
-						propertiesBuilder.Add(_builder.ToString(firstIndexOf + 1, lastIndexOf - firstIndexOf - 1));
+						// Setting the Length here will remove the ".Method" part from the string builder since we want to add "SomePath" only.
+						propertyBuilder.Length = lastIndexOf;
+						AddIfNotStaticAndClear(propertyBuilder, propertiesBuilder);
+					}
+					else
+					{
+						propertyBuilder.Clear();
 					}
 
 					_builder.Append('(');
 					for (int i = 0; i < invocation.Arguments.Length; i++)
 					{
-						var oldLength = _builder.Length;
 						var argument = invocation.Arguments[i];
 						if (argument is XBindPathArgument pathArgument)
 						{
-							BuildPath(pathArgument.Path);
-							var newLength = _builder.Length;
-							var argPath = _builder.ToString(oldLength, newLength - oldLength);
-							if (argPath.StartsWith($"{_contextName}.", StringComparison.Ordinal))
-							{
-								propertiesBuilder.Add(argPath.Substring(_contextName.Length + 1));
-							}
-							else if (argPath is not ("null" or "true" or "false")) // TODO: Consider special nodes for these.
-							{
-								propertiesBuilder.Add(argPath);
-							}
+							BuildPath(pathArgument.Path, propertyBuilder);
+							AddIfNotStaticAndClear(propertyBuilder, propertiesBuilder);
 						}
 						else if (argument is XBindLiteralArgument literalArgument)
 						{
@@ -154,14 +138,11 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 				}
 				else if (root is XBindPath path)
 				{
-					BuildPath(path);
-					var csharpPath = _builder.ToString();
-					if (csharpPath.StartsWith($"{_contextName}.", StringComparison.Ordinal))
-					{
-						return (csharpPath, ImmutableArray.Create(csharpPath.Substring(_contextName.Length + 1)));
-					}
+					BuildPath(path, propertyBuilder);
+					AddIfNotStaticAndClear(propertyBuilder, propertiesBuilder);
 
-					return (csharpPath, ImmutableArray.Create(csharpPath));
+					var csharpPath = _builder.ToString();
+					return (csharpPath, propertiesBuilder.ToImmutableArray());
 				}
 				else if (root is null)
 				{
@@ -171,17 +152,19 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 				throw new Exception("Unexpected XBindRoot");
 			}
 
-			public void BuildPath(XBindPath path)
+			public void BuildPath(XBindPath path, StringBuilder? propertyBuilder)
 			{
 				if (path is XBindMemberAccess memberAccess)
 				{
-					BuildPath(memberAccess.Path);
+					BuildPath(memberAccess.Path, propertyBuilder);
+					propertyBuilder?.Append($".{memberAccess.Identifier.IdentifierText}");
 					_builder.Append('.');
 					_builder.Append(memberAccess.Identifier.IdentifierText);
 				}
 				else if (path is XBindIndexerAccess indexerAccess)
 				{
-					BuildPath(indexerAccess.Path);
+					BuildPath(indexerAccess.Path, propertyBuilder);
+					propertyBuilder?.Append($"[{indexerAccess.Index}]");
 					_builder.Append('[');
 					_builder.Append(indexerAccess.Index);
 					_builder.Append(']');
@@ -196,6 +179,7 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 					}
 
 					_builder.Append(identifier.IdentifierText);
+					propertyBuilder?.Append(identifier.IdentifierText);
 				}
 				else if (path is XBindAttachedPropertyAccess attachedPropertyAccess)
 				{
@@ -205,19 +189,19 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 					}
 					else
 					{
-						BuildPath(attachedPropertyAccess.PropertyClass);
+						BuildPath(attachedPropertyAccess.PropertyClass, propertyBuilder: null);
 					}
 
 					_builder.Append(".Get");
 					_builder.Append(attachedPropertyAccess.PropertyName.IdentifierText);
 					_builder.Append('(');
-					BuildPath(attachedPropertyAccess.Member);
+					BuildPath(attachedPropertyAccess.Member, propertyBuilder: null);
 					_builder.Append(')');
 				}
 				else if (path is XBindParenthesizedExpression parenthesizedExpression)
 				{
 					_builder.Append('(');
-					BuildPath(parenthesizedExpression.Expression);
+					BuildPath(parenthesizedExpression.Expression, propertyBuilder);
 					_builder.Append(')');
 
 					if (parenthesizedExpression.IsPathlessCast)
@@ -228,9 +212,9 @@ namespace Uno.UI.SourceGenerators.XamlGenerator.Utils
 				else if (path is XBindCast xBindCast)
 				{
 					_builder.Append('(');
-					BuildPath(xBindCast.Type);
+					BuildPath(xBindCast.Type, propertyBuilder: null);
 					_builder.Append(')');
-					BuildPath(xBindCast.Expression);
+					BuildPath(xBindCast.Expression, propertyBuilder);
 				}
 				else
 				{

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/When_xBind_With_Cast.xaml
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/When_xBind_With_Cast.xaml
@@ -1,0 +1,30 @@
+﻿<Page
+    x:Class="Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml.Controls.When_xBind_With_Cast"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml.Controls"
+	xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+    mc:Ignorable="d"
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
+
+	<muxc:NavigationView x:Name="NavView"
+                         PaneDisplayMode="Top" IsBackButtonVisible="Collapsed">
+		<muxc:NavigationView.MenuItems>
+			<muxc:NavigationViewItem x:Name="ItemHelp" x:FieldModifier="public">
+				<muxc:NavigationViewItem.Icon>
+					<FontIcon FontFamily="{ThemeResource SymbolThemeFontFamily}" Glyph="&#xE9CE;" />
+				</muxc:NavigationViewItem.Icon>
+			</muxc:NavigationViewItem>
+			<muxc:NavigationViewItem x:Name="ItemOther" x:FieldModifier="public" Content="Other" Icon="Calendar"
+                                     IsSelected="True"/>
+			<muxc:NavigationViewItem x:Name="ItemOther2" x:FieldModifier="public" Content="Other 2">
+				<muxc:NavigationViewItem.Icon>
+					<FontIcon FontFamily="{ThemeResource SymbolThemeFontFamily}" Glyph="&#xF272;" />
+				</muxc:NavigationViewItem.Icon>
+			</muxc:NavigationViewItem>
+		</muxc:NavigationView.MenuItems>
+		<TextBlock x:Name="tb" x:FieldModifier="public" Text="{x:Bind ((muxc:NavigationViewItem)NavView.SelectedItem).Name, Mode=OneWay}" />
+	</muxc:NavigationView>
+</Page>

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/When_xBind_With_Cast.xaml.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Controls/When_xBind_With_Cast.xaml.cs
@@ -1,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices.WindowsRuntime;
+using Windows.Foundation;
+using Windows.Foundation.Collections;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
+
+// The Blank Page item template is documented at https://go.microsoft.com/fwlink/?LinkId=234238
+
+namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml.Controls;
+/// <summary>
+/// An empty page that can be used on its own or navigated to within a Frame.
+/// </summary>
+public sealed partial class When_xBind_With_Cast : Page
+{
+	public When_xBind_With_Cast()
+	{
+		this.InitializeComponent();
+	}
+}

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Given_xBind.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Given_xBind.cs
@@ -9,6 +9,9 @@ namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml;
 public class Given_xBind
 {
 	[TestMethod]
+#if __IOS__
+	[Ignore("Fails on iOS")]
+#endif
 	public async Task When_xBind_With_Cast()
 	{
 		var SUT = new When_xBind_With_Cast();

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Given_xBind.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml/Given_xBind.cs
@@ -1,0 +1,29 @@
+﻿using Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml.Controls;
+using Private.Infrastructure;
+using System.Threading.Tasks;
+
+namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml;
+
+[TestClass]
+[RunsOnUIThread]
+public class Given_xBind
+{
+	[TestMethod]
+	public async Task When_xBind_With_Cast()
+	{
+		var SUT = new When_xBind_With_Cast();
+		TestServices.WindowHelper.WindowContent = SUT;
+		await TestServices.WindowHelper.WaitForLoaded(SUT);
+
+		Assert.AreEqual("ItemOther", SUT.tb.Text);
+
+		SUT.ItemHelp.IsSelected = true;
+		Assert.AreEqual("ItemHelp", SUT.tb.Text);
+
+		SUT.ItemOther2.IsSelected = true;
+		Assert.AreEqual("ItemOther2", SUT.tb.Text);
+
+		SUT.ItemOther.IsSelected = true;
+		Assert.AreEqual("ItemOther", SUT.tb.Text);
+	}
+}

--- a/src/Uno.UI/UI/Xaml/Style/mergedstyles.xaml
+++ b/src/Uno.UI/UI/Xaml/Style/mergedstyles.xaml
@@ -14538,12 +14538,12 @@
     <Setter Property="ScrollViewer.HorizontalScrollBarVisibility" Value="Auto" />
     <Setter Property="ScrollViewer.VerticalScrollMode" Value="Auto" />
     <Setter Property="ScrollViewer.VerticalScrollBarVisibility" Value="Auto" />
-    <win:Setter Property="ScrollViewer.ZoomMode" Value="Disabled" />
+    <Setter Property="ScrollViewer.ZoomMode" Value="Disabled" />
     <Setter Property="Template">
       <Setter.Value>
         <ControlTemplate TargetType="FlyoutPresenter">
           <Border Background="{TemplateBinding Background}" BackgroundSizing="OuterBorderEdge" BorderBrush="{TemplateBinding BorderBrush}" BorderThickness="{TemplateBinding BorderThickness}" CornerRadius="{TemplateBinding CornerRadius}" Padding="{ThemeResource FlyoutBorderThemePadding}">
-            <ScrollViewer x:Name="ScrollViewer" not_win:ZoomMode="Disabled" win:ZoomMode="{TemplateBinding ScrollViewer.ZoomMode}" HorizontalScrollMode="{TemplateBinding ScrollViewer.HorizontalScrollMode}" HorizontalScrollBarVisibility="{TemplateBinding ScrollViewer.HorizontalScrollBarVisibility}" VerticalScrollMode="{TemplateBinding ScrollViewer.VerticalScrollMode}" VerticalScrollBarVisibility="{TemplateBinding ScrollViewer.VerticalScrollBarVisibility}" AutomationProperties.AccessibilityView="Raw">
+            <ScrollViewer x:Name="ScrollViewer" ZoomMode="{TemplateBinding ScrollViewer.ZoomMode}" HorizontalScrollMode="{TemplateBinding ScrollViewer.HorizontalScrollMode}" HorizontalScrollBarVisibility="{TemplateBinding ScrollViewer.HorizontalScrollBarVisibility}" VerticalScrollMode="{TemplateBinding ScrollViewer.VerticalScrollMode}" VerticalScrollBarVisibility="{TemplateBinding ScrollViewer.VerticalScrollBarVisibility}" AutomationProperties.AccessibilityView="Raw">
               <ContentPresenter Content="{TemplateBinding Content}" ContentTemplate="{TemplateBinding ContentTemplate}" ContentTransitions="{TemplateBinding ContentTransitions}" Margin="{TemplateBinding Padding}" HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}" VerticalAlignment="{TemplateBinding VerticalContentAlignment}" />
             </ScrollViewer>
           </Border>

